dotCMS/src/main/java/com/dotcms/rest/api/v1/template/TemplateResource.java

```diff
@@ -0,0 +1,752 @@
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.PermissionAPI;
+import com.dotmarketing.business.RoleAPI;
+import com.dotmarketing.business.VersionableAPI;
+import com.dotmarketing.business.web.HostWebAPI;
+import com.dotmarketing.business.web.WebAPILocator;
+import com.dotmarketing.exception.DoesNotExistException;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.factories.WebAssetFactory;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.folders.business.FolderAPI;
+import com.dotmarketing.portlets.htmlpageasset.business.HTMLPageAssetAPI;
+import com.dotmarketing.portlets.htmlpageasset.model.HTMLPageAsset;
+import com.dotmarketing.portlets.templates.business.TemplateAPI;
+import com.dotmarketing.portlets.templates.business.TemplateConstants;
+import com.dotmarketing.portlets.templates.design.util.DesignTemplateUtil;
+import com.dotmarketing.portlets.templates.model.Template;
+import com.dotmarketing.util.ActivityLogger;
+import com.dotmarketing.util.InodeUtils;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.PageMode;
+import com.dotmarketing.util.UtilMethods;
+import com.dotmarketing.util.WebKeys;
+import com.google.common.annotations.VisibleForTesting;
+import com.liferay.portal.language.LanguageException;
+import com.liferay.portal.language.LanguageUtil;
+import com.liferay.portal.model.User;
+import com.liferay.util.servlet.SessionDialogMessage;
+import io.vavr.control.Try;
+import org.glassfish.jersey.server.JSONP;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.dotcms.util.CollectionsUtils.map;
+import static com.dotmarketing.business.PermissionAPI.PERMISSION_READ;
+import static com.dotmarketing.business.PermissionAPI.PERMISSION_WRITE;
+
+/**
+ * CRUD of Templates
+ * @author jsanca
+ */
+@Path("/v1/templates")
+public class TemplateResource {
+
+    private final PaginationUtil paginationUtil;
+    private final WebResource    webResource;
+    private final TemplateAPI    templateAPI;
+    private final VersionableAPI versionableAPI;
+    private final FolderAPI      folderAPI;
+    private final HostWebAPI     hostWebAPI;
+    private final PermissionAPI  permissionAPI;
+    private final RoleAPI        roleAPI;
+    private final HTMLPageAssetAPI htmlPageAssetAPI;
```

Comment:
Codacy found an issue: [Perhaps 'htmlPageAssetAPI' could be replaced by a local variable.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6042990)